src/test/java/org/fiware/vc/it/model/Application.java

```diff
@@ -0,0 +1,128 @@
+		List<NameValuePair> params = URLEncodedUtils.parse(URI.create(locationHeader), Charset.forName("UTF-8"));
+		SameDeviceParams sameDeviceParams = new SameDeviceParams();
+		params.forEach(p -> {
+			switch (p.getName()) {
+				case "response_type" -> sameDeviceParams.setResponseType(p.getValue());
+				case "response_mode" -> sameDeviceParams.setResponseMode(p.getValue());
+				case "client_id" -> sameDeviceParams.setClientId(p.getValue());
+				case "redirect_uri" -> sameDeviceParams.setRedirectUri(p.getValue());
+				case "state" -> sameDeviceParams.setState(p.getValue());
+				case "nonce" -> sameDeviceParams.setNonce(p.getValue());
+				case "scope" -> sameDeviceParams.setScope(p.getValue());
+				default -> log.warn("Received an unknown parameter: {}", p.getName());
+			}
+		});
+
+		assertEquals("vp_token", sameDeviceParams.getResponseType(), "Currently, only vp_token is supported.");
+		assertEquals("direct_post", sameDeviceParams.getResponseMode(), "Currently, only direct_post is supported.");
+		assertEquals(applicationConfig.verifierDid(), sameDeviceParams.getClientId(),
+				"The expected participant should have initiated the flow.");
+		assertNotNull(sameDeviceParams.getRedirectUri(), "A redirect_uri should have been received.");
+		assertNotNull(sameDeviceParams.getState(), "The verifier should have creadet a state.");
+
+		return sameDeviceParams;
+	}
+
+	public void exchangeCodeForJWT(AuthResponseParams authResponseParams) throws Exception {
+		Map<String, String> tokenRequestFormData = Map.of(
+
+				"grant_type", "authorization_code",
+				"code", authResponseParams.getCode(),
+				// we did not set a redirect_path, thus in same device we will end up where it began.
+				"redirect_uri", applicationConfig.applicationUrl() + "/");
+
+		HttpRequest jwtRequest = HttpRequest.newBuilder()
+				.uri(URI.create(
+						String.format("%s%s", applicationConfig.verifierAddress(), applicationConfig.tokenPath())))
+				.POST(HttpRequest.BodyPublishers.ofString(getFormDataAsString(tokenRequestFormData)))
+				.header("Content-Type", MediaType.APPLICATION_FORM_URLENCODED)
+				.build();
+		HttpResponse<String> tokenResponse = HTTP_CLIENT.send(jwtRequest, HttpResponse.BodyHandlers.ofString());
+		assertEquals(HttpStatus.SC_OK, tokenResponse.statusCode(), "A token should have been returned.");
+		TokenResponse tr = OBJECT_MAPPER.readValue(tokenResponse.body(), TokenResponse.class);
+
+		jwt = tr.getAccessToken();
+	}
+
+	public void accessBackend() throws Exception {
+
+		HttpRequest orionRequest = HttpRequest.newBuilder()
+				.uri(URI.create(String.format("%s/ngsi-ld/v1/entities/urn:ngsi-ld:DELIVERYORDER:1",
+						applicationConfig.securedBackend())))
+				.GET()
+				.header("Authorization", "Bearer " + jwt)
+				.build();
+		HttpResponse<String> orionResponse = HTTP_CLIENT.send(orionRequest, HttpResponse.BodyHandlers.ofString());
+		assertEquals(404, orionResponse.statusCode(), "The request should have been allowed.");
```

Comment:
I understand that you basically test !=401/403 but 404 seems weird (eg cover up missing service endpoint) :sweat_smile: Could you instead issue a count or some other simple call that returns a proper result?

Reply:
Yeah, I should do that...:/